Pointer, Offset and ByteCount abstractions in mmalloc

gkk-stripe opened the pull request:
Introudce Pointer, Offset and ByteCount abstractions to mmalloc
implementation and its APIs. The motivation is the not use naked Ints
and Longs for distinct concepts like capacity, memory offset or memory
pointer. By introducing wrapper classes, we can track these things more
explicitly and separately.

The direct motivation for this change was an integer overflow bug that
was caused by mixing Ints and Longs in an expression that was computing
a pointer based of an offset.

This commit fixes two integer overflow bugs that I found.

gkk-stripe committed:
Pointer, Offset and ByteCount abstractions in mmalloc

Introudce Pointer, Offset and ByteCount abstractions to mmalloc
implementation and its APIs. The motivation is the not use naked Ints
and Longs for distinct concepts like capacity, memory offset or memory
pointer. By introducing wrapper classes, we can track these things more
explicitly and separately.

The direct motivation for this change was an integer overflow bug that
was caused by mixing Ints and Longs in an expression that was computing
a pointer based of an offset.

This commit fixes two integer overflow bugs that I found.

gkk-stripe commented:
r? @peterl-stripe

peterl-stripe commented on src/main/scala/co/teapot/mmalloc/ByteBufferBasedBitSet.scala:
```diff
@@ -53,7 +53,7 @@ private[mmalloc] class ByteBufferBasedBitSet(
     for (groupI <- (startGroupI until groupCount).view ++ (0 until startGroupI)) {
       if (doesGroupContainBit(groupI, set)) {
-        val group = data.getInt(pointer + groupI * 4)
+        val group = data.getInt(pointer + Offset.ints(groupI))
```
Nice, that does look at a lot cleaner.

peterl-stripe commented on src/main/scala/co/teapot/mmalloc/LargeMappedByteBuffer.scala:
```diff
@@ -70,37 +70,41 @@ class MMapByteBuffer(f: File) extends LargeMappedByteBuffer {
+  override def getInt(index: Pointer): Int = memory.getInt(index.raw)
+
+  def putInt(index: Pointer, v: Int, forceToDisk: Boolean = false): Unit = {
+    ensureSize(index + Offset(3))
```
Would `Offset.bytes(3)` be more clear here, since you use `Offset.ints` above?

gkk-stripe replied:
Yes, good idea.

peterl-stripe commented on src/main/scala/co/teapot/mmalloc/MemoryMappedAllocator.scala:
```diff
@@ -22,20 +22,86 @@ import it.unimi.dsi.fastutil.longs.LongArrayList
+  def toByteCount: ByteCount = ByteCount(raw)
+}
+object Offset {
+  def fromBlocks(blockCount: Int, blockSize: ByteCount): Offset =
```
In line with `ints` and `longs`, would `blocks` be a more consistent method name here?

peterl-stripe commented on src/main/scala/co/teapot/mmalloc/PieceBlock.scala:
```diff
@@ -83,63 +84,66 @@ private[mmalloc] class PieceBlock(val pointer: Long, mmAllocator: MemoryMappedAl
+  def pieceSize(pointer: Pointer, data: LargeMappedByteBuffer): ByteCount =
+    ByteCount(data.getInt(pointer + PieceSizeOffset))
+  def setPieceSize(pointer: Pointer, data: LargeMappedByteBuffer, pieceSize: ByteCount, syncWrite: Boolean): Unit = {
+//    require(pointer.raw + PieceSizeOffset > 0, "Possible integer overflow found")
```
Is this comment still useful, or just left from debugging?

gkk-stripe replied:
I'll keep it. I think it's useful.

peterl-stripe commented on src/main/scala/co/teapot/mmalloc/PieceBlock.scala:
```diff
@@ -83,63 +84,66 @@ private[mmalloc] class PieceBlock(val pointer: Long, mmAllocator: MemoryMappedAl
     ByteBufferBasedBitSet.initializeWith1s(pointer + BitsetOffset, data, pieceCount)
     if (syncWrite) {
-      data.syncToDisk(pointer, BitsetOffset + pieceCount / 8)
+      data.syncToDisk(pointer, ByteCount(BitsetOffset.raw + pieceCount / 8))
```
You don't have to, but this might be clearer as something like
`ByteCount(BitsetOffset + Offset.bits(pieceCount))`

gkk-stripe replied:
👍

peterl-stripe commented on src/main/scala/co/teapot/tempest/graph/MemMappedDynamicUnidirectionalGraph.scala:
```diff
@@ -59,27 +59,35 @@ class MemMappedDynamicUnidirectionalGraph(allocator: MemoryMappedAllocator,
-  def maxNodeId: Int = maxNodeIdLong.toInt
+  def maxNodeId: Int = {
+    val result = maxNodeIdLong
+    require(result < Int.MaxValue)
```
Nice check.

peterl-stripe commented on src/main/scala/co/teapot/tempest/graph/MemMappedDynamicUnidirectionalGraph.scala:
```diff
@@ -59,27 +59,35 @@ class MemMappedDynamicUnidirectionalGraph(allocator: MemoryMappedAllocator,
   }
 
-  private def nodeCapacity: Long = data.getLong(dataPointer + NodeCapacityOffset)
+  private def nodeCapacity: Int = {
```
Why change`nodeCapacity` to be an `Int`? I think it's possible we'll want between 2^31 and 2^32 nodes again in the future.

gkk-stripe replied:
I'd hold this off for the focused effort. For now, it's better to assume that 2^31 is the max id.

peterl-stripe commented on src/main/scala/co/teapot/tempest/graph/MemMappedDynamicUnidirectionalGraph.scala:
```diff
@@ -59,27 +59,35 @@ class MemMappedDynamicUnidirectionalGraph(allocator: MemoryMappedAllocator,
+  private def nodePointer(id: Int): Pointer = {
     require(Integer.toUnsignedLong(id) < nodeCapacity, s"id ${Integer.toUnsignedLong(id)} >= nodeCapacity $nodeCapacity")
-    nodeArrayPointer + BytesPerNode * Integer.toUnsignedLong(id)
+    nodeArrayPointer + Offset.fromBlocks(id, BytesPerNode)
```
`Integer.toUnsignedLong` was to support having between 2^31 and 2^32 nodes.  It was pretty hacky though, so I'm fine with this PR removing `Integer.toUnsignedLong`.

gkk-stripe replied:
👍

peterl-stripe reviewed (approved):
Thanks, this looks great for eliminating pointer arithmetic mistakes!  It must have take a while to go through and find all the places where pointers are used.  I made some minor comments, but everything looks correct to me, so it's up to you which you want to respond to.
LGTM

gkk-stripe committed:
Fix the LargeMappedByteBufferSpec test

There were some code refactoring changes before and I missed fixing this
test.

gkk-stripe committed:
Address review feedback

Changes:
  - introduce Offset.bytes and Offset.bits
  - rename Offset.fromBlocks to Offset.blocks

gkk-stripe commented:
Review feedback addressed.
